lib/formatters/basic_formatter.rb

```diff
@@ -89,6 +89,23 @@ def print()
       puts format_value(@msg_content_hashed ? StringUtils.sha1_hash(@message.body) : @message.body)
     end # print()
 
+    # Escapes characters which Python's eval() cannot load
+    # that is esp. \0, \r, \n. Use a range, to be safe.
+    def self.escape_chars(msg_string)
+      if msg_string.nil?
+        nil
+      else
+        msg_string.each_codepoint.map do |s|
+          # \\\\ because of python eval escaping
+          if s < 32 or s > 126
+            format('\\\\u%04x', s)
+          else
+            s.chr
+          end
+        end.join
```

Comment:
@alanconway hi, not sure if you read GH notifications. just in case you knew how to do this nicer in ruby... I tried regexp with negative lookahead on [[:print:]] character group... but did not get anywhere

Reply:
If you just want to quote non-printable characters:

irb> puts "null=\0 nl=\n shamrock=\u2618".inspect[1..-2]
null=\u0000 nl=\n shamrock=☘

If you want to quote all non-ascii characters:

irb(main):052:0> puts "null=\0 nl=\n shamrock=\u2618".force_encoding("ascii").inspect[1..-2]
null=\x00 nl=\n shamrock=\xE2\x98\x98

String.inspect returns a source-code representation of the string, including surrounding quotes, [1..-2] means to drop the first and last characters - which are the quotes.

Reply:
@rkubis welcome back ;) I'd go with the .inspect solution, I think. Also, I believe that unicode need not be quoted for our purposes, so i'll leave shamrocks be.

Reply:
solution without inspect #50